### PR TITLE
Add status to OCS resources.

### PR DESCRIPTION
**Pull request by sacharya**
This commit adds status conditions to OCS resources and provides a framework to aggregate component statuses. As component operators are integrated, their statuses have to be evaluated and turned into OCS status. This uses [custom-resource-status](https://github.com/openshift/custom-resource-status), which provides some level of
standardization and best practices to managing the status of custom
resources.

### CODE REVIEW

File: deploy/crds/ocs_v1alpha1_ocsinitialization_crd.yaml [hunk 1]
```diff
@@ -31,8 +31,39 @@ spec:
           type: object
         status:
           properties:
-            errorMessage:
-              type: string
+            conditions:
+              description: Conditions describes the state of the HyperConverged resource.
```
Comment (jarrpa) [inline deploy/crds/ocs_v1alpha1_ocsinitialization_crd.yaml]: Remove `HyperConverged`

File: deploy/crds/ocs_v1alpha1_storagecluster_crd.yaml [hunk 2]
```diff
@@ -24,6 +32,8 @@ spec:
             instanceType:
               type: string
             manageNodes:
+              description: 'Important: Run "operator-sdk generate k8s" to regenerate
```
Comment (jarrpa) [inline deploy/crds/ocs_v1alpha1_storagecluster_crd.yaml]: Remove this description.

File: deploy/crds/ocs_v1alpha1_storagecluster_crd.yaml [hunk 3]
```diff
@@ -62,6 +72,40 @@ spec:
               type: array
           type: object
         status:
+          properties:
+            conditions:
+              description: Conditions describes the state of the HyperConverged resource.
```
Comment (jarrpa) [inline deploy/crds/ocs_v1alpha1_storagecluster_crd.yaml]: Remove `HyperConverged`.

File: pkg/controller/storagecluster/reconcile.go [hunk 4]
```diff
@@ -66,20 +146,35 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1alpha1.StorageClust
...
-	} else if err == nil {
-		// Update the CephCluster if it is not in the desired state
-		if !reflect.DeepEqual(cephCluster.Spec, found.Spec) {
-			reqLogger.Info("Updating spec for CephCluster")
-			found.Spec = cephCluster.Spec
-			err = r.client.Update(context.TODO(), found)
-		}
+		return r.client.Create(context.TODO(), cephCluster)
+	}
+	if err != nil {
```
Comment (jarrpa) [inline pkg/controller/storagecluster/reconcile.go]: Nit: I see why you would move the following logic out of its own stanza, but I don't like this solution. Please combine this if statement with the previous one, i.e. something like:
```
if err != nil {
  if NotFound() {
    logInfo()
    err = Create()
  }
  return err
}
```

File: deploy/crds/ocs_v1alpha1_storagecluster_crd.yaml [hunk 5]
```diff
@@ -62,6 +70,40 @@ spec:
               type: array
           type: object
         status:
+          properties:
+            conditions:
+              description: Conditions describes the state of the StorageCluster resource.
+                +patchMergeKey=type +patchStrategy=merge
```
Comment (kshlm) [inline deploy/crds/ocs_v1alpha1_storagecluster_crd.yaml]: Was this meant to be included? Doesn't seem to be belong.

File: pkg/controller/storagecluster/reconcile.go [hunk 6]
```diff
@@ -119,3 +215,85 @@ func newCephCluster(sc *ocsv1alpha1.StorageCluster) *rookCephv1.CephCluster {
 		},
 	}
 }
+
+// setProgressingCondition sets the ProgressingCondition to True and other conditions to
+// false or Unknown. Used when we are just starting to reconcile, and there are no existing
+// conditions.
+func setProgressingCondition(conditions *[]conditionsv1.Condition, reason string, message string) {
```
Comment (kshlm) [inline pkg/controller/storagecluster/reconcile.go]: This is being duplicated from the ocsinitialization controller. Can you put this and the functions below, into a new package, and call them that way.

**Comment by kshlm** [comment]
The golint job failed on generated files. It shouldn't be checking them.

@jarrpa You can ignore the lint failure and manually merge this. I'll fix the CI job tomorrow.

**Comment by jarrpa** [comment]
Looks like golint failed on one of the generated files... maybe update the version of operator-sdk being used? @kshlm Is it possible something changed in the version or configuration of our golist test?

**Comment by jarrpa** [comment]
@kshlm Will do.

**Comment by jarrpa** [comment]
Looks like I can't force the merge because the master branch is protected. It'll have to wait until you fix the test.

**Comment by kshlm** [comment]
@jarrpa You should be able to override the status with `/override ci/prow/golint`. https://prow.k8s.io/command-help#override

**Pull request merged**